File: windows/SeamlessClip/Protocol/Framing.cs
using System.Buffers.Binary;

namespace SeamlessClip.Protocol;

/// <summary>Length-prefixed frames: uint32 big-endian length, then payload.</summary>
public static class Framing
{
    public static async Task<byte[]> ReadFrameAsync(Stream stream, int maxLength, CancellationToken ct)
    {
        var header = new byte[4];
        await stream.ReadExactlyAsync(header, ct).ConfigureAwait(false);
        int length = BinaryPrimitives.ReadInt32BigEndian(header);
        if (length < 0 || length > maxLength)
            throw new InvalidDataException($"Frame length {length} exceeds limit {maxLength}");

        var payload = new byte[length];
        await stream.ReadExactlyAsync(payload, ct).ConfigureAwait(false);
        return payload;
    }

    public static async Task WriteFrameAsync(Stream stream, ReadOnlyMemory<byte> payload, CancellationToken ct)
    {
        var buffer = new byte[4 + payload.Length];
        BinaryPrimitives.WriteInt32BigEndian(buffer, payload.Length);
        payload.CopyTo(buffer.AsMemory(4));
        await stream.WriteAsync(buffer, ct).ConfigureAwait(false);
        await stream.FlushAsync(ct).ConfigureAwait(false);
    }
}

File: android/app/src/main/java/app/seamlessclip/protocol/Framing.kt
package app.seamlessclip.protocol

import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException

/** Length-prefixed frames: uint32 big-endian length, then payload. */
object Framing {
    fun read(input: DataInputStream, maxLength: Int): ByteArray {
        val length = input.readInt()
        if (length < 0 || length > maxLength) throw IOException("Frame length $length exceeds limit $maxLength")
        val payload = ByteArray(length)
        input.readFully(payload)
        return payload
    }

    fun write(output: DataOutputStream, payload: ByteArray) {
        output.writeInt(payload.size)
        output.write(payload)
        output.flush()
    }
}
